src/primitives/line/styled.rs

```diff
@@ -64,26 +169,87 @@ impl<C: PixelColor> StyledDrawable<PrimitiveStyle<C>> for Line {
         D: DrawTarget<Color = C>,
     {
-        target.draw_iter(StyledPixelsIterator::new(self, style))
+        let dot_size = style.stroke_width as i32;
+        if dot_size <= 0 {
+            return Ok(());
+        }
+
+        let Some(stroke_color) = style.effective_stroke_color() else {
+            return Ok(());
+        };
+
+        if style.stroke_style == StrokeStyle::Dotted && dot_size > 1 {
+            let line = if dot_size % 2 == 0 {
+                // When drawing a dotted rectangle border, the distance between the endpoint dots
+                // is longer by one pixel when the dot size is even.
+                // So that the dotted rectangle border matches 4 lines drawn in clockwise order,
+                // the line is extended by one pixel when the dot size is even.
+                extend_line_by_one_unit(self)
+            } else {
+                *self
+            };
+            // Draw circles along the line.
+            let mut length = line.delta().length_squared().integer_sqrt();
+            // The gaps between dots ideally have the same size as the dots
+            // If `dot_size <= 3`, only positive error is allowed,
+            // otherwise both positive and negative error are allowed.
+            if dot_size > 3 {
+                length += dot_size;
+            }
+            // The 2 endpoint dots take half the space of a regular dot.
+            let nb_dots_desired = length / (2 * dot_size) + 1;
+            let dotted_line_points =
+                DottedLinePoints::new(&line.translate(-line.start), nb_dots_desired);
+            let dot_style = PrimitiveStyle::with_fill(stroke_color);
+            // Improve the positioning of the dots.
+            let start = start_dot_offset(self, dot_size);
+
+            if dot_size > 3 {
+                draw_dotted_line(
+                    &Circle::with_center(start, dot_size as u32),
+                    &dotted_line_points,
+                    &dot_style,
+                    target,
+                )
+            } else {
+                draw_dotted_line(
+                    &Rectangle::with_center(start, Size::new_equal(dot_size as u32)),
+                    &dotted_line_points,
+                    &dot_style,
+                    target,
+                )
+            }
+        } else if style.stroke_style == StrokeStyle::Dotted && style.stroke_width == 1 {
+            target.draw_iter(StyledPixelsIterator::new(self, style).step_by(2))
+        } else {
+            target.draw_iter(StyledPixelsIterator::new(self, style))
+        }
     }
 }
 
 impl<C: PixelColor> StyledDimensions<PrimitiveStyle<C>> for Line {
     fn styled_bounding_box(&self, style: &PrimitiveStyle<C>) -> Rectangle {
-        let (l, r) = self.extents(style.stroke_width, StrokeOffset::None);
+        if style.stroke_style == StrokeStyle::Dotted {
+            let min = self.start.component_min(self.end);
+            let max = self.start.component_max(self.end);
```

Comment:
This is redundant, because `Rectangle::with_corners` also determines the min and max coordinates.